### PR TITLE
Fixed distinct composite type class proc borrowing

### LINKED ISSUE #7552: Static + distinct: Cannot borrow when static is involved
Test case:

```Nim
type
  Foo = distinct seq[int]

  Bar[N: static[int]] = distinct seq[int]


proc newSeq(s: var Foo, n: Natural) {.borrow.} # works

proc newSeq(s: var Bar, n: Natural) {.borrow.} # Error: no symbol to borrow from found
# proc newSeq[N: static[int]](s: var Bar[N], n: Natural) {.borrow.} # Same error
```

### PR DESCRIPTION
**Pull request by beef331**
Closes: https://github.com/nim-lang/Nim/issues/7552
Also indirectly tests for a no longer existing issue of: https://github.com/nim-lang/Nim/issues/7553

### CODE REVIEW

File: compiler/semcall.nim [hunk 1]
```diff
@@ -717,8 +726,12 @@ proc searchForBorrowProc(c: PContext, startScope: PScope, fn: PSym): PSym =
     var resolved = semOverloadedCall(c, call, call, filter, {})
     if resolved != nil:
       result = resolved[0].sym
+      while sfBorrow in result.flags:
+        # We iterate over all borrowed procs.
+        # This ensures we always have an instantiated proc to call
+        result = result.ast[bodyPos].sym
```
Comment (Araq) [inline compiler/semcall.nim]: No, the underlying proc was not called here, the borrowed one was. Think about "find all usages" -- you want to list the usages of the borrowed proc, not the usages of the underlying proc. `transf.nim` does the step you added here.

Comment (beef331) [reply compiler/semcall.nim]: Without that logic the following errors due to Baz's proc pointing to Bar's proc which isn't instantiated. Erroring with `Error: internal error: expr: proc not init $`. Presently transf doesnt handle this, but if it should I can move the logic there.
```nim
type
  Foo = distinct seq[int]
  Bar[N: static[int]] = distinct seq[int]
  Baz = distinct Bar[10]
  
proc `$`(s: Foo): string {.borrow.}
proc `$`(s: Bar): string {.borrow.}
proc `$`(s: Baz): string {.borrow.}
```

Comment (Araq) [reply compiler/semcall.nim]: Yes, patch `transf` instead.

File: compiler/semcall.nim [hunk 2]
```diff
@@ -703,7 +703,16 @@ proc searchForBorrowProc(c: PContext, startScope: PScope, fn: PSym): PSym =
...
   for i in 1..<fn.typ.n.len:
     let param = fn.typ.n[i]
-    let t = skipTypes(param.typ, abstractVar-{tyTypeDesc, tyDistinct})
+    const desiredTypes = abstractVar + {tyCompositeTypeClass} - {tyTypeDesc, tyDistinct}
+    #[.
+      # We only want the type not any modifiers such as `ptr`, `var`, `ref` ...
+      # tyCompositeTypeClass is here for
+      # when using something like:
+      type Foo[T] = distinct int
+      proc `$`(f: Foo): string {.borrow.}
```
Comment (Araq) [inline compiler/semcall.nim:712]: In a follow-up PR, please update the spec to mention that this is valid.

**Pull request merged**